File: src/main/scala/com/vc/medium/L238.scala
package com.vc.medium

object L238 {
  def productExceptSelf(arr: Array[Int]): Array[Int] = {
    val n = arr.length
    val res = new Array[Int](n)
    res(0) = 1
    (1 until n).foreach(i => {
      res(i) = arr(i - 1) * res(i - 1)
    })

    var runningProduct = 1
    (n - 1 to 0 by -1).foreach(i => {
      res(i) *= runningProduct
      runningProduct *= arr(i)
    })
    res
  }
}

File: src/main/java/com/vc/medium/L238.java
package com.vc.medium;

class L238 {
    public int[] productExceptSelf(int[] nums) {
        if(nums == null || nums.length == 0) return new int[0];

        int n = nums.length;
        int[] left = new int[n];
        int[] right = new int[n];

        int runningProduct = 1;
        for(int i = 0; i < n; i++) {
            left[i] = runningProduct;
            runningProduct *= nums[i];
        }

        runningProduct = 1;
        for(int i = n - 1; i >= 0; i--) {
            right[i] = runningProduct;
            runningProduct *= nums[i];
        }

        int[] res = new int[n];
        for(int i = 0; i < res.length; i++) {
            res[i] = left[i] * right[i];
        }

        return res;
    }
}
